src/torch/README.md

```diff
@@ -0,0 +1,123 @@
+    4. Saves relevant information to the `ctx` if gradients are needed.
+
+*   **Backward Pass (VJP):**
+    1.  Receives the upstream gradient (`grad_output`), which is typically a tensor containing the scalar `1` (since the output of the forward pass is a scalar, and `∂L/∂(-f(x)) = 1` if `L = -f(x)`).
+    2.  Gets the gradient of the *negative* objective function using `sirf_obj_func.get_gradient()`, which computes `-∇f(x)`.
+    3.  Converts the SIRF gradient to a PyTorch tensor.
+    4.  Returns the gradient multiplied by `grad_output`.
+
+    **More on the VJP:** Let `f(x)` be the SIRF objective function. The forward pass computes `-f(x)`.  The backward pass needs to compute `∂L/∂x`, where `L` is the overall loss function (which, in the simplest case, *is* the output of the forward pass, `-f(x)`). Using the chain rule:
+    `∂L/∂x = (∂L/∂(-f(x))) * (∂(-f(x))/∂x) = grad_output * (-∇f(x))`.  Since the forward pass output is a scalar, the Jacobian is simply the gradient `∇f(x)`.  The upstream gradient `grad_output` represents `∂L/∂(-f(x))`, and `sirf_obj_func.get_gradient()` provides `-∇f(x)`.  The multiplication `grad_output * (-∇f(x))` correctly computes the VJP.
+
+### `_ObjectiveFunctionGradient` (Forward and Backward Passes)
+
+*   **Forward Pass (JVP with v=1):**
+    1.  Converts the input PyTorch tensor to a SIRF object.
+    2.  Computes the *gradient* of the *negative* SIRF objective function using `sirf_obj_func.get_gradient()`, which returns `-∇f(x)`.  This is equivalent to computing a JVP where the input vector `v` is implicitly `1` (because we are taking the full gradient of the *scalar* output of the objective function). The result is a vector, the gradient.
+    3.  Returns the (negative) gradient as a PyTorch tensor.
+
+*   **Backward Pass (Hessian-Vector Product - HVP):**
+    1.  Receives the upstream gradient (`grad_output`), which now represents a *vector* (not a scalar).  This vector corresponds to `∂L/∂g(x)`, where `g(x) = -∇f(x)` is the *output* of the forward pass (the negative gradient of the objective function).
+    2.  Converts `grad_output` to a SIRF object.
+    3.  Computes the Hessian-vector product (HVP) using `sirf_obj_func.multiply_with_Hessian()`.  This method takes two arguments: the current estimate (`sirf_image`, where the Hessian is evaluated) and the input vector (`sirf_grad`, which is `grad_output` converted to a SIRF object).
+    4. Returns the (negative) HVP.
+
+    **More on the HVP:**  Let `f(x)` be the SIRF objective function, and let `g(x) = -∇f(x)` be the output of the forward pass. We want to compute `∂L/∂x`, where `L` is the overall loss. The backward pass receives `v = grad_output = ∂L/∂g(x)`.  The Hessian of `f(x)` is `H = ∇²f(x)`. We compute the HVP: `H|x * v`, where `H|x` denotes the Hessian evaluated at point x. Applying the chain rule:
```

Comment:
@KrisThielemans does this seem a better explanation? Also I added GD with sirf acquistion model. There are other use cases in the use_case.py